projects/driver_controls/src/direction_fsm.c

```diff
@@ -0,0 +1,96 @@
+#include "direction_fsm.h"
+#include "input_event.h"
+#include "event_arbiter.h"
+#include "log.h"
+
+// Direction selector FSM state definitions
+
+FSM_DECLARE_STATE(state_neutral);
+FSM_DECLARE_STATE(state_forward);
+FSM_DECLARE_STATE(state_reverse);
+
+// Direction selector FSM transition table definitions
+
+FSM_STATE_TRANSITION(state_neutral) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_DRIVE, state_forward);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_REVERSE, state_reverse);
+}
+
+FSM_STATE_TRANSITION(state_forward) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL, state_neutral);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_REVERSE, state_reverse);
+}
+
+FSM_STATE_TRANSITION(state_reverse) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_DRIVE, state_forward);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL, state_neutral);
+}
+
+// Direction selector FSM arbiter functions
+
+static bool prv_check_neutral(const Event *e) {
+  // The car must not be able to move while in the neutral state
+  return !(e->id == INPUT_EVENT_PEDAL_COAST || e->id == INPUT_EVENT_PEDAL_PRESSED);
+}
+
+static bool prv_check_forward(const Event *e) {
+  // Powering off while the car is moving can be dangerous. As a result, powering off is
+  // forbidden while in forward gear
+  return (e->id != INPUT_EVENT_POWER);
+}
+
+static bool prv_check_reverse(const Event *e) {
+  // Same as forward, except that cruise control is forbidden in reverse for obvious reasons
+  return !(e->id == INPUT_EVENT_POWER || e->id == INPUT_EVENT_CRUISE_CONTROL);
+}
+
+// Direction selector FSM output functions
+
+static void prv_state_neutral(FSM *fsm, const Event *e, void *context) {
+  EventArbiterCheck *event_check = fsm->context;
+  *event_check = prv_check_neutral;
+
+  InputEventData data;
+  data.raw = e->data;
```

Comment:
To be more explicit, I think it would be good to set `data.components.data = e->data` instead of setting raw directly. Also add a comment to make it clear that data is limited to 12 bits.